File: sdk/rust/src/artifact/gh.rs
use crate::{
    api::artifact::ArtifactSystem::{Aarch64Darwin, Aarch64Linux, X8664Darwin, X8664Linux},
    artifact::{step, ArtifactBuilder, ArtifactSourceBuilder},
    context::ConfigContext,
};
use anyhow::{bail, Result};
use indoc::formatdoc;

pub async fn build(context: &mut ConfigContext) -> Result<String> {
    let name = "gh";
    let system = context.get_system();
    match system {
        Aarch64Darwin | Aarch64Linux | X8664Darwin | X8664Linux => {}
        _ => bail!("unsupported {name} system: {}", system.as_str_name()),
    };

    let source_target = match system {
        Aarch64Darwin => "macOS_arm64",
        Aarch64Linux => "linux_arm64",
        X8664Darwin => "macOS_amd64",
        X8664Linux => "linux_amd64",
        _ => bail!("unsupported {name} system: {}", system.as_str_name()),
    };

    let source_extension = match system {
        Aarch64Darwin | X8664Darwin => "zip",
        Aarch64Linux | X8664Linux => "tar.gz",
        _ => bail!("unsupported {name} system: {}", system.as_str_name()),
    };

    let source_version = "2.69.0";
    let source_path = format!("https://github.com/cli/cli/releases/download/v{source_version}/gh_{source_version}_{source_target}.{source_extension}");
    let source = ArtifactSourceBuilder::new(name, source_path.as_str()).build();

    let step_script = formatdoc! {"
        mkdir -pv \"$VORPAL_OUTPUT/bin\"

        cp -prv \"source/{name}/gh_{source_version}_{source_target}/bin/gh\" \"$VORPAL_OUTPUT/bin/gh\"

        chmod +x \"$VORPAL_OUTPUT/bin/gh\"",
    };

    let step = step::shell(context, vec![], vec![], step_script, vec![]).await?;
    let systems = vec![Aarch64Darwin, Aarch64Linux, X8664Darwin, X8664Linux];

    ArtifactBuilder::new(name, vec![step], systems)
        .with_aliases(vec![format!("{name}:{source_version}")])
        .with_sources(vec![source])
        .build(context)
        .await
}

File: sdk/python/src/vorpal_sdk/artifact/gh.py
"""GitHub CLI (gh) artifact (prebuilt release archive).

Mirrors ``sdk/typescript/src/artifact/gh.js``.
"""

from __future__ import annotations

from typing import TYPE_CHECKING

from vorpal_sdk.artifact import Artifact, ArtifactSource
from vorpal_sdk.step import shell
from vorpal_sdk.system import SYSTEMS, get_system_str

if TYPE_CHECKING:
    from vorpal_sdk.context import ConfigContext

DEFAULT_GH_VERSION = "2.87.3"


class Gh:
    """Builder for the GitHub CLI artifact."""

    def build(self, context: ConfigContext) -> str:
        name = "gh"
        system = context.get_system()
        system_str = get_system_str(system)

        if system_str == "aarch64-darwin":
            source_target = "macOS_arm64"
            source_extension = "zip"
        elif system_str == "aarch64-linux":
            source_target = "linux_arm64"
            source_extension = "tar.gz"
        elif system_str == "x86_64-darwin":
            source_target = "macOS_amd64"
            source_extension = "zip"
        elif system_str == "x86_64-linux":
            source_target = "linux_amd64"
            source_extension = "tar.gz"
        else:
            raise ValueError(f"unsupported {name} system: {system}")

        source_version = DEFAULT_GH_VERSION
        source_path = (
            f"https://sdk.vorpal.build/source/"
            f"gh_{source_version}_{source_target}.{source_extension}"
        )

        source = ArtifactSource(name, source_path).build()

        cp_line = (
            f'cp -pr "source/{name}/gh_{source_version}_{source_target}/bin/gh" '
            f'"$VORPAL_OUTPUT/bin/gh"'
        )
        step_script = (
            'mkdir -p "$VORPAL_OUTPUT/bin"\n\n'
            f"{cp_line}\n\n"
            'chmod +x "$VORPAL_OUTPUT/bin/gh"'
        )

        steps = [shell(context, [], [], step_script, [])]

        return (
            Artifact(name, steps, SYSTEMS)
            .with_aliases([f"{name}:{source_version}"])
            .with_sources([source])
            .build(context)
        )
